magicgui/function_gui.py

```diff
@@ -0,0 +1,286 @@
+
+        return_type = self._return_annotation
+        if return_type:
+            for callback in _type2callback(return_type):
+                callback(self, value, return_type)
+        self.called(value=value)
+        return value
+
+    def __repr__(self) -> str:
+        """Return string representation of instance."""
+        fname = f"{self._function.__module__}.{self._function.__name__}"
+        return f"<FunctionGui {fname}{self.to_signature()}>"
+
+    @property
+    def result_name(self) -> str:
+        """Return a name that can be used for the result of this magicfunction."""
+        return self._result_name or (self._function.__name__ + " result")
+
+    @result_name.setter
+    def result_name(self, value: str):
+        """Set the result name of this MagicGui widget."""
+        self._result_name = value
+
+    def Gui(self, show=False):
+        """Create a widget instance [DEPRECATED]."""
+        warnings.warn(
+            "\n\nCreating a widget instance with `my_function.Gui()` is deprecated,\n"
+            "the magicgui decorator now returns a widget instance directly, so you\n"
+            "should simply use the function itself as a magicgui widget, or call\n"
+            "`my_function.show(run=True)` to run the application.\n"
+            "In a future version, the `Gui` attribute will be removed.\n",
+            FutureWarning,
+        )
+        if show:
+            self.show()
+        return self
+
+
+# ==================   magicgui decorator   ===================================
+
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+@overload
+def magicgui(function: F, **k) -> FunctionGui:  # noqa: D103
+    ...
+
+
+@overload
+def magicgui(function=None, **k) -> Callable[[F], FunctionGui]:  # noqa: D103
+    ...
+
+
+def magicgui(
+    function: Optional[F] = None,
+    *,
+    orientation: str = "horizontal",
+    labels: bool = True,
+    call_button: Union[bool, str] = False,
+    auto_call: bool = False,
+    result_widget: bool = False,
+    app: AppRef = None,
+    **param_options: dict,
+):
+    """Create a FunctionGui class for ``function`` and add it as an attribute ``Gui``.
+
+    Parameters
+    ----------
+    function : Callable, optional
+        The function to decorate.  Optional to allow bare decorator with optional
+        arguments. by default None
+    orientation : str, optional
+        The type of layout to use. Must be one of {'horizontal', 'vertical',
+        'form', 'grid'}, by default "horizontal"
+    labels : bool
+        Whether labels are shown in the widget. by default True
+    call_button : bool or str, optional
+        If True, create an additional button that calls the original function when
+        clicked.  If a ``str``, set the button text. by default False
+    auto_call : bool, optional
+        If True, changing any parameter in either the GUI or the widget attributes
+        will call the original function with the current settings. by default False
```

Comment:
Looks like `app` doc string is missing (not trying to be pedantic, but just flagging in case accidentally missed)

Reply:
good catch! thanks